deepchem/models/torch_model.py

```diff
@@ -0,0 +1,1167 @@
+  ways.  For models that have only a single output and use a
+  standard loss function, you can simply provide a
+  dc.models.losses.Loss object.  This defines the loss for each
+  sample or sample/task pair.  The result is automatically
+  multiplied by the weights and averaged over the batch.  Any
+  additional losses computed by model layers, such as weight
+  decay penalties, are also added.
+
+  For more complicated cases, you can instead provide a function
+  that directly computes the total loss.  It must be of the form
+  f(outputs, labels, weights), taking the list of outputs from
+  the model, the expected values, and any weight matrices.  It
+  should return a scalar equal to the value of the loss function
+  for the batch.  No additional processing is done to the
+  result; it is up to you to do any weighting, averaging, adding
+  of penalty terms, etc.
+
+  You can optionally provide an output_types argument, which
+  describes how to interpret the model's outputs.  This should
+  be a list of strings, one for each output. You can use an
+  arbitrary output_type for a output, but some output_types are
+  special and will undergo extra processing:
+
+  - 'prediction': This is a normal output, and will be returned by predict().
+    If output types are not specified, all outputs are assumed
+    to be of this type.
+
+  - 'loss': This output will be used in place of the normal
+    outputs for computing the loss function.  For example,
+    models that output probability distributions usually do it
+    by computing unbounded numbers (the logits), then passing
+    them through a softmax function to turn them into
+    probabilities.  When computing the cross entropy, it is more
+    numerically stable to use the logits directly rather than
+    the probabilities.  You can do this by having the model
+    produce both probabilities and logits as outputs, then
+    specifying output_types=['prediction', 'loss'].  When
+    predict() is called, only the first output (the
+    probabilities) will be returned.  But during training, it is
+    the second output (the logits) that will be passed to the
+    loss function.
+
+  - 'variance': This output is used for estimating the
+    uncertainty in another output.  To create a model that can
+    estimate uncertainty, there must be the same number of
+    'prediction' and 'variance' outputs.  Each variance output
+    must have the same shape as the corresponding prediction
+    output, and each element is an estimate of the variance in
+    the corresponding prediction.  Also be aware that if a model
+    supports uncertainty, it MUST use dropout on every layer,
+    and dropout most be enabled during uncertainty prediction.
+    Otherwise, the uncertainties it computes will be inaccurate.
+
+  - other: Arbitrary output_types can be used to extract outputs
+    produced by the model, but will have no additional
+    processing performed.
```

Comment:
It might be useful to add an examples section to the docstring demonstrating how to use TorchModel on a simple toy model

Reply:
Added!  (Also for KerasModel.)